src/app/notebook/notebook-helpers.ts

```diff
@@ -0,0 +1,47 @@
+import { MimeBundle } from './notebook-types';
+
+// A list of the mime types we will display, ordered from "best" to "worst".
+export const supportedMimeTypes: string[] = ['image/png', 'image/webp', 'image/jpeg', 'image/gif', 'text/html', 'text/json', 'text/plain'];
+
+export function selectBestFromMimeBundle(mimeBundle: MimeBundle): { mimeType: string; data: string } {
+  for (const mimeType of supportedMimeTypes) {
+    const data = mimeBundle[mimeType];
+    if (data != undefined) {
+      return { mimeType: mimeType, data: join(data) };
+    }
+  }
+  return undefined;
+}
+
+export function join(value: any): string {
+  if (value == undefined) {
+    return '';
+  }
+  if (Array.isArray(value)) {
+    return value.join('');
+  }
+  return String(value);
+}
+
+// The below escape() implementation is adapted from mustache.js
+// https://github.com/janl/mustache.js/blob/972fd2b27a036888acfcb60d6119317744fac7ee/mustache.js#L60
+export const escapeCharToEntity = {
```

Comment:
Not my favorite to see we are now manually maintaining this given angular has this feature somewhere but seems simple enough. (ngSanitize?)

Reply:
Yeah, I wasn't able to find an alternative that gave me more confidence than the known good implementation that I horked (the included mustache.js adaptation).  You'd think it'd be accessible via Angular, but I couldn't find it, and maybe, in general, they want to discourage this kind of thing.

Reply:
You're not invoking this from outside the file, so you don't need to export it.